### ISSUE TITLE
RPMs no longer support PHP 7.1 on RHEL

### ISSUE THREAD

**Issue by XolphinMaarten** [Icinga/icinga-packaging#69]
I have been using Remi's PHP 7.1 packages (php-7.1.12-2.el7.remi.x86_64) together with Icinga2 packages for a long time, but I am unable to upgrade to the latest version. Yum is now suggesting to update to rh-php71-php* packages, which I do not have.

Please change the spec file to allow for other PHP packages as well.

See https://rpms.remirepo.net/enterprise/7/php71/x86_64/repoview/ for the PHP packages involved.

**Comment by Reamer** [comment]
Some problem on my site

**Comment by dnsmichi** [comment]
SCL is a requirement for Icinga Web 2 v2.5 and officially supported by RHEL. Remi is a community repository, still their packages are also found in SCLO.

**Comment by XolphinMaarten** [comment]
Why would SCL be a requirement for Icinga Web? Wouldn't any PHP 5.6+ version do? I prefer to have just a single PHP 7 version on my system, and not use the ugly /opt/rh/ installation method as SCL uses. I think the Remi repository is much more used as well, because more people prefer it. I think it would make sense to support this as well.

**Comment by dnsmichi** [comment]
Do you know a possible way of RPM supporting either one package dependency or the other? That's the main problem here, AFAIK this does not work.

**Comment by XolphinMaarten** [comment]
No, sorry, I don't. I think the easiest is to remove the requirement for the PHP version in the spec file, or create a separate RPM with a normal PHP 7 dependency instead of requiring the SCL one.

**Comment by lippserd** [comment]
We will think about a solution for this problem.

**Comment by SimonHoenscheid** [comment]
We also had to change. Problems with SCO is the context

**Comment by lazyfrosch** [comment]
Problem is basically that SCL is the only official way of going onto a higher PHP.

While external repos can offer packages that replace some in the base distribution, SCL tries to avoid that.

Other than that, RPM doesn't allow alternative dependencies.

We could only fix that by:
* Shipping our own PHP (what we definitely won't do)
* Offering another repo based on Remi's packages

Both are not really an option for Enterprise users, which are targeted by our builds here. So we use the "official" way.

You can remove all php-* packages from the system, unless you need them for other applications.

**Comment by TwizzyDizzy** [comment]
Mhhh... OK, I see the predicament here. Yet, I think, there is a third option as described above by @XolphinMaarten: remove the PHP dependency.

I also see another option: by default create two package versions... one with the dependency to SCL PHP, one without any PHP dependency.

Aren't those valid options, too? No dependency is an obvious bummer, 'cause it impedes the "it's working directly after installing" experience of users.

Cheers
Thomas

**Comment by lazyfrosch** [comment]
As @lippserd said, we will discuss this and give you feedback here

**Comment by XolphinMaarten** [comment]
After discussing this on the CentOS IRC chat, avij suggested the following hack:

Create two intermediate RPM's, a icingaweb2-php7-native which depends on the normal PHP RPM (the PHP 7.1 from the Remi repository for example), and a icingaweb2-php7-scl which depends on the RedHat provided SCL PHP (rh-php71). Both RPM's provide icingaweb2-php7. The main RPM can then be modified to require icingaweb2-php7, and then you can use either one of them.

**Comment by JochenFriedrich** [comment]
With IUS, it's the same.

https://ius.io/IUSvsSCL/

**Comment by sbeee** [comment]
We're on Oracle Linux (v7.4)...dont't ask...we've no choice. :(  The Problem is that Oracle's SCL still is at version 2.4 (php 7.0). It takes weeks or months until Oracle will release a newer version of this repository. :-/

Now i've the same problem as others here. I can't install icingaweb2 because of dependency errors (rh-php71-php).

As Oracle Linux User i don't have rh-php71-php, just rh-php70-php.

**Comment by lazyfrosch** [comment]
CentOS SCL repositories are public and should work on every EPEL

**Comment by wols** [comment]
First, I know, packaging and distribute is your hard work - thanks for all.
But the decision for `rh-php71-*` has bad effects.
```
---> Package php-Icinga.noarch 0:2.5.0-1.el7.icinga will be an update
--> Processing Dependency: rh-php71-php-common >= 5.6.0 for package: php-Icinga-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-intl for package: php-Icinga-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-xml for package: php-Icinga-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-gd for package: php-Icinga-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-pdo for package: php-Icinga-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-mbstring for package: php-Icinga-2.5.0-1.el7.icinga.noarch
```
```
---> Package icingaweb2.noarch 0:2.5.0-1.el7.icinga will be an update
--> Processing Dependency: icingacli = 2.5.0-1.el7.icinga for package: icingaweb2-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-common >= 5.6.0 for package: icingaweb2-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-fpm >= 5.6.0 for package: icingaweb2-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-pgsql for package: icingaweb2-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-mysqlnd for package: icingaweb2-2.5.0-1.el7.icinga.noarch
--> Processing Dependency: rh-php71-php-ldap for package: icingaweb2-2.5.0-1.el7.icinga.noarch
```
Not all installations have unlimited internet and/or repository access! Releases for mirroring in safety-critical environments, this can take months of process :-1: 
I do not see any possibility because only the one package to champion their own php.

Open source should also offer the possibility "to have the choice".
Unfortunately, at the moment I only have the possibility for `yum versionlock`. That is very unsatisfactory for me. But I have to live with the version provided by the OS and mirrored by my environment.

**Comment by ctrlaltca** [comment]
I just wanted to add that the same problem applies to other RHEL derivatives, in my case Amazon Linux (AWS), that already supports different php version packages (eg php56-*, php70-*, php71-*) without the need of external repos.
Until now using the http://packages.icinga.com/epel/6/release/ repo worked flawlessly, but the hard dependency on the SCL repo is now making it impossible to update without dirty hacks (rpm -i manually the packages) or having to maintain two php installations on the same box.

**Comment by lippserd** [comment]
Hi guys,

First, thanks to all for sharing your thoughts in this discussion.

Our main focus is to provide well tested packages for all major platforms. That means that we do not directly support RHEL derivates like Amazon and Oracle Linux. If their SCL repo is out of date, you have the option to use the SCL from Red Hat or CentOS.

Our choice to depend on the SCL should be obvious:
* RHEL and CentOS do not directly offer PHP 5.6+
* Vendor libraries and dependencies which are not under our control should be actively and directly supported by the vendor
* The SCL in this regard is the only repository supported by Red Hat

We discussed the option to provide different PHP dependency packages. This would of course require us to build and test more packages. Further, we would have to test those packages with all the prominent community repos like IUS, Remi and Webtactics. Also it is not guaranteed that this approach just works and the continuous maintenance efforts are not foreseeable.

Honestly, for the only benefit of "I don't like SCL please use my community or derivate repo" we are not making this happen.

Since the SPEC file is [publicly available](https://github.com/Icinga/icinga-packaging/blob/rpm/release/icingaweb2/icingaweb2.spec#L19,L25) you may build the packages yourselves with the PHP dependency you like.
If you need an older version of `rh-php`, just change the marked lines from the above linked SPEC file. If you do not want use SCL, comment the line which defines the SCL prefix. Make sure to use `#%`as prefix because percentage signs (%) have to be escaped separately in SPEC files.

Best regards,
Eric

**Comment by TwizzyDizzy** [comment]
While I am not particularly happy with this decision ('cause it means work), I do understand support the reasoning... looks like moving to SCL then.

Cheers
Thomas

**Comment by ctrlaltca** [comment]
Just an reminder on how to workaround this problem for amazon linux and other "illegitimate" child distros:
```
# grab and install manually the centos-scl-rh repo
# Note: this is the version for distros based on RHEL6
rpm -i http://mirror.centos.org/centos/6/extras/x86_64/Packages/centos-release-scl-rh-2-3.el6.centos.noarch.rpm
```
This will install 179MB of unused stuff in `/opt/rh`, that seems to be a chroot env for php7.0 and a chroot env for python2.7.
Then icingaweb can be installed/updated normally, and it will work with the system/installed php as long as it respect the minimum version requirement (currently php 5.6).
Remember to disable the repo when finished:
```
yum-config-manager --disable centos-sclo-rh
```
If you need to update the package in the future because the requirements change, you can temporary enable the repos:
```
yum --enablerepo epel,centos-scl-rh update
```

**Comment by JochenFriedrich** [comment]
No, you still use the SCL stuff as the path is hardcoded in /usr/bin/icingacli.

I now use this with IUS stuff:
rh-php71-php-common.spec:

Summary:        A fake package to avoid crappy SCL stuff
Name:           rh-php71-php-common
Version:        5.6.0
Release:        1
License:        GPL
Group:          Applications/System
Requires:       php-common >= 5.6.0
Requires:       php-cgi >= 5.6.0
BuildArch:      noarch
BuildRoot:      %{tmpdir}/%{name}-%{version}-build
Provides:       rh-php71-php-cli = 5.6.0
Provides:       rh-php71-php-ldap = 5.6.0
Provides:       rh-php71-php-intl = 5.6.0
Provides:       rh-php71-php-mysqlnd = 5.6.0
Provides:       rh-php71-php-pgsql = 5.6.0
Provides:       rh-php71-php-gd = 5.6.0
Provides:       rh-php71-php-xml = 5.6.0
Provides:       rh-php71-php-fpm = 5.6.0
Provides:       rh-php71-php-pdo = 5.6.0
Provides:       rh-php71-php-mbstring = 5.6.0

%description
A fake package to avoid crappy SCL stuff

%prep

%install
mkdir -p $RPM_BUILD_ROOT/opt/rh/rh-php71/root/usr/bin
ln -s /usr/bin/php $RPM_BUILD_ROOT/opt/rh/rh-php71/root/usr/bin/php

%clean
rm -fr $RPM_BUILD_ROOT

%files
%defattr(644,root,root,755)
%attr(755,root,root) /opt/rh/rh-php71/root/usr/bin/php

**Comment by ctrlaltca** [comment]
> ctrlaltca, you still use the SCL stuff as the path is hardcoded in /usr/bin/icingacli.

As far as i understand, your statement is true as long as i run `icingacli web serve` to use icinga's internal web server (based on php-fpm, i suppose).
Instead, i just configured an apache virtualhost providing web access to `/usr/share/icingaweb2/public` with the same exact config i previously used for icingaweb2 2.4 and it seems to run just fine.

**Comment by dnsmichi** [comment]
If the macro `php_scl` is not set, `php_bin` won't be enabled either. That one takes care of modifying the path in the icingacli php script.

https://github.com/Icinga/icinga-packaging/blob/rpm/release/icingaweb2/icingaweb2.spec#L33

Modify the spec file as you need.

**Comment by JochenFriedrich** [comment]
Modifying and repackaging icinga2 requires a manual step each time icinga2 is updated. This will prevent an automated software distribution process like spacewalk to keep a huge infrastructure up to date.
That's why I prefer a simple one time solution and keep reposync running against the official icinga2 packages.

**Comment by bcogel** [comment]
It's not only the PHP version. I have to switch to apache 2.4 from software collection. This breaks our setup on several systems. 

- I have to rewrite the complete configuration/configuration management for all these systems. 
- Logging doesn't go to /var/log (this is a separate filessystem on our systems). Instead /var/opt is used, which is in /var. With possibly not enough filesspace. 

I have to redesign the complete system layout and test the setup. This has far to much impact for a simple software update!!!

You don't have to support different community repos. You only have to support clean requirements for php >= 5.6.0 (Requires: php-common >= 5.6.0, Requires: php-cgi >= 5.6.0). Each properly build RPM will give the information what it provides so dnf/yum/rpm can decide if the requirements are met. 
And if you have to modify the php path in the icingacli script, this could be done in postinstall. Or it can be a simple configuration option in /etc/icingaweb2/config.ini. 

But the real problem is that Red Hat doesn't provide the required information in their packages:

rpm -q --whatprovides php
mod_php71u-7.1.12-1.ius.el6.x86_64

Ok.... the software collection RPM doesn't provide PHP.....X-(

rpm -qa|grep '.*php.*common.*'
rh-php70-php-common-7.0.10-2.el6.x86_64
php71u-common-7.1.12-1.ius.el6.x86_64

rpm -q --provides mod_php71u-7.1.12-1.ius.el6.x86_64
...
php = 7.1.12-1.ius.el6
...

rpm -q --provides php71u-common-7.1.12-1.ius.el6.x86_64
...
php(language) = 7.1.12
...
php-core = 7.1.12
...
php71u-core = 7.1.12
...
php-common = 7.1.12-1.ius.el6
...

rpm -q --provides rh-php70-php-common-7.0.10-2.el6.x86_64
...
rh-php70-php(language) = 7.0.10
...
rh-php70-php-core = 7.0.10
...

They don't use simple php capabilities. Instead they use rh-php70.. tags. This is the source of the problem... And because one cannot use OR and AND in requirements, there is only the solution with a dummy package like the one above.

What about providing such a package in the icinga repo. It can also provide a config for icingacli with the correct php_bin path (mostly /usr/bin/php, at least with IUS). In icingacli look for the config file (or do a 'which php'). If it doesn't exit, fall back to the SCL path. Of course you need two versions: for RHEL6 and RHEL7. But this can be done in the SPEC file.

If an admin wants to use his own php version, he only has to include the dummy package in the installation. Or... if it is done right, this package will be installed automatically to fulfill the requirements when no SCL Package is already installed but a PHP 7.x is available. And if no PHP 7.x is installed, but one of the configured repos provides a correctly build version, this version will be installed because of the requirements in the dummy package.

**Comment by bcogel** [comment]
If one of the community versions doesn't work properly, the admin has the option to switch to another repo or use SCL. But most of those versions have been thoroughly tested. This is the reason why they are used and still exist.

**Comment by lazyfrosch** [comment]
@bcogel 
> It's not only the PHP version. I have to switch to apache 2.4 from software collection. This breaks our setup on several systems.

This is not true, unless you explicitly want to use mod_php, when you use FPM like we recommend then, you can use any webserver.

> But the real problem is that Red Hat doesn't provide the required information in their packages:

Because SCLs are designed not to interfere with other packages inside the base distribution, because they have to provide compatibility.

RPM in general has not good handling for versioned `Provides`, this is a major problem for SLES - where we have to conflict with older PHP, so it won't install them (RPM ignores the version via Provides there)

**Comment by annunaki2k2** [comment]
Just wanted to add my own remarks on this as we've just come up against this "feature" too.

It's a pretty major headache for us. We use Remi's PHP 7.1 repo and packages extensively throughout our (puppet controlled) infrastructure because it's always up-to-date, always reliable, it's fully integrated into our puppet modules and manifests, and (especially as) it doesn't need us to set additional paths when running non-interactive login shells (my biggest bug bear with any stuff /opt/rh).

We don't have our own RPM repository (or honestly, the time either) in which to host and roll our own versions of the these packages.

I feel we really shouldn't have to re-write half our puppet manifests to cope with the fact that icingaweb2 will now only use one fixed source of php packages - it should use what is given already on the system, and if it doesn't work, well then that's my problem (after-all, I choose to run it outside of the requirements), but at least let _that_ be my problem rather than preventing us from installing the software at all.

Please re-open this issue and release a version free of the SCL requirements so we can at the very least get this installed.

**Comment by bcogel** [comment]
I'm now using a dummy package on RHEL6.  

Btw.: The config in /etc/httpd/conf.d/icingaweb2.conf should not only check for the apache version. There should also be a check wether the needed module is loaded. You assume that FPM is used. 

I had to modify the default configuration because I cannot use FPM at the moment. 

php-fpm from IUS requires httpd24 from IUS.

**Comment by bcogel** [comment]
With the dummy package almost everything works fine. Except when I want to trigger a recheck of a service. Then I'm getting an error:

icinga2: Got invalid JSON response from the Icinga 2 API: Syntax error: '<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>400 Bad Request</title>
</head><body>
<h1>Bad Request</h1>
<p>Your browser sent a request that this server could not understand.<br />
</p>
<hr>
<address>Apache/2.2.15 (Red Hat) Server at XXXX Port 443</address>
</body></html>
'.

#0 /usr/share/icingaweb2/modules/monitoring/application/forms/Command/Object/CheckNowCommandForm.php(70): Icinga\Module\Monitoring\Command\Transport\CommandTransport->send(Object(Icinga\Module\Monitoring\Command\Object\ScheduleServiceCheckCommand))
#1 /usr/share/php/Icinga/Web/Form.php(1154): Icinga\Module\Monitoring\Forms\Command\Object\CheckNowCommandForm->onSuccess()
#2 /usr/share/icingaweb2/modules/monitoring/library/Monitoring/Web/Controller/MonitoredObjectController.php(264): Icinga\Web\Form->handleRequest()
#3 /usr/share/icingaweb2/modules/monitoring/library/Monitoring/Web/Controller/MonitoredObjectController.php(59): Icinga\Module\Monitoring\Web\Controller\MonitoredObjectController->setupQuickActionForms()
#4 /usr/share/icingaweb2/modules/monitoring/application/controllers/ServiceController.php(66): Icinga\Module\Monitoring\Web\Controller\MonitoredObjectController->showAction()
#5 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Monitoring\Controllers\ServiceController->showAction()
#6 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch('showAction')
#7 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#8 /usr/share/php/Icinga/Application/Web.php(407): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#9 /usr/share/php/Icinga/Application/webrouter.php(104): Icinga\Application\Web->dispatch()
#10 /usr/share/icingaweb2/public/index.php(4): require_once('/usr/share/php/...')
#11 {main}

I don't know if this happens because icingacli depends on FPM or if this has another reason. I'm on my own with over 200 linux systems and other tasks and I need a working monitoring solution to keep all up and runnig. Nobody here has any experince with PHP-FPM. And I simply don't have the time and resources to make all the required changes for FPM or SCL. I'm not keen on 'always the latest and greatest'. I'm an admin and not a developer..... So until there is any other solution I've downgraded to icingaweb 2.4.2.

**Comment by bcogel** [comment]
Hmm... we're using commandtransport via api. May be this is broken in the new version. When we temporarily switch to command pipe transport, recheck works as expected. 

We're using commandtransport via api to synchronize comments, acknowledgements and downtimes between clustered icinga instances and their webfrontends.

**Comment by lippserd** [comment]
We'll think about intermediate RPMs for our next major release which is not yet scheduled though.
It is not guaranteed that we'll provide them but we'll try it at least.

@bcogel Please open a new issue for the API related part.
